# 作業三: 分類器的原理與評比實驗
## 資料來源: 來自 3 個產區，178 瓶葡萄酒，含 13 種葡萄酒成分
## 目標: 
## 計畫執行這篇講義描述的分類器比較，即採用三種分類器分別對三組資料進行分類學習與測試。其中分類器包括： 
## 1.多元羅吉斯回歸 2.支援向量機 3.神經網路
## 影像資料處理: 
## 1.原始資料 2.進行PCA主成分分析
## 分類方法: 
- ## Logistic Regression
- ## SVM
- ## Neural Network
### 姓名: 鄭欣莉
### 學號: 410877039

## 標準化後原始資料+羅吉斯回歸

In [2]:
import numpy as np 
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

# Read data
df = pd.read_excel('wine.xlsx')
X = np.array(df.iloc[:, :-1]) # 排 除 最 後 一 欄 標 籤
y = np.array(df.iloc[:, -1])

# Split data into training and testing data
test_size = 0.30 #多少比較好
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=test_size) #分完train,test資料之後再標準化比較好
#標準化
scaler = StandardScaler()
X_train_ = scaler.fit_transform(X_train)
X_test_ = scaler.fit_transform(X_test)

- solver = '’lbfgs’'

In [3]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report

def clf_LR(solver):
    opts = dict(tol = 1e-6, max_iter = int(1e6), verbose=1) #tol=tolerance
    clf_LR = LogisticRegression(solver = solver, **opts)
    clf_LR.fit(X_train_, y_train)
    y_pred = clf_LR.predict(X_test_)
    # 測 試 資 料 之 準 確 率 回 報
    print(f"{accuracy_score(y_test, y_pred):.2%}\n")
    print(f"{clf_LR.score(X_test_, y_test):.2%}\n")
    print(classification_report(y_test, y_pred))

In [4]:
solver = 'lbfgs' # ’lbfgs’ is the default 
clf_LR(solver)

96.30%

96.30%

              precision    recall  f1-score   support

           1       0.93      1.00      0.97        14
           2       0.95      0.95      0.95        20
           3       1.00      0.95      0.97        20

    accuracy                           0.96        54
   macro avg       0.96      0.97      0.96        54
weighted avg       0.96      0.96      0.96        54



[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.0s finished


- solver = 'liblinear'

In [5]:
clf_LR(solver = 'liblinear')

[LibLinear]98.15%

98.15%

              precision    recall  f1-score   support

           1       1.00      1.00      1.00        14
           2       0.95      1.00      0.98        20
           3       1.00      0.95      0.97        20

    accuracy                           0.98        54
   macro avg       0.98      0.98      0.98        54
weighted avg       0.98      0.98      0.98        54



- solver = 'newton-cg'

In [6]:
clf_LR(solver = 'newton-cg')

96.30%

96.30%

              precision    recall  f1-score   support

           1       0.93      1.00      0.97        14
           2       0.95      0.95      0.95        20
           3       1.00      0.95      0.97        20

    accuracy                           0.96        54
   macro avg       0.96      0.97      0.96        54
weighted avg       0.96      0.96      0.96        54



[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.0s finished


## 結論:
### 1. Classification Report 的衡量指標個別代表的意思:
- ### 精確率(Precision) 為預測為真的樣本有幾個預測正確
- ### 召回率(Recall) 為事實為真的樣本中有幾個是預測正確的
- ### F1 score 為精確率和召回率的調和平均數
### 2. 以不同solver的第一筆資料為例:
- ### solver = 'lbfgs' 跟 'newton-cg'預測為真且預測正確的比例為93%，solver = 'liblinear'為100%
- ### 每個 solver 在事實為真的樣本中預測正確率為百分之百
- ### solver = 'lbfgs' 跟 'newton-cg'精確率和召回率的調和平均為97%，solver = 'liblinear'為100%
### 3. 標準化過後的AT&T資料，solver = 'lbfgs' 跟 'newton-cg' 表現均差不多，準確率均在90%以上，solver = 'liblinear' 準確率最高98%，各個 solver 表現不錯

## PCA主成分分析(成分比例採0.8)+羅吉斯回歸

- solver = 'lbfgs'

In [9]:
from sklearn.decomposition import PCA

def PCA_LR(n_components,solver):
    pca = PCA(n_components = n_components).fit(X_train_) 
    Z_train = pca.transform(X_train_)
    Z_test = pca.transform(X_test_)
    opts = dict(tol = 1e-4, max_iter = int(1e6), verbose=1)
    clf_PCA = LogisticRegression(solver = solver, **opts)
    clf_PCA.fit(Z_train, y_train)
    y_pred = clf_PCA.predict(Z_test)
    print(f"{clf_PCA.score(Z_test, y_test):.2%}\n")
    print(classification_report(y_test, y_pred))

In [10]:
solver = 'lbfgs'
PCA_LR(n_components = 0.8, solver = solver)

92.59%

              precision    recall  f1-score   support

           1       0.88      1.00      0.93        14
           2       0.95      0.90      0.92        20
           3       0.95      0.90      0.92        20

    accuracy                           0.93        54
   macro avg       0.92      0.93      0.93        54
weighted avg       0.93      0.93      0.93        54



[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.0s finished


- solver = 'liblinear'

In [12]:
solver = 'liblinear'
PCA_LR(n_components = 0.8, solver = solver)

[LibLinear]92.59%

              precision    recall  f1-score   support

           1       0.88      1.00      0.93        14
           2       0.95      0.90      0.92        20
           3       0.95      0.90      0.92        20

    accuracy                           0.93        54
   macro avg       0.92      0.93      0.93        54
weighted avg       0.93      0.93      0.93        54



- solver = 'newton−cg'

In [11]:
solver = 'newton-cg'
PCA_LR(n_components = 0.8, solver = solver)

92.59%

              precision    recall  f1-score   support

           1       0.88      1.00      0.93        14
           2       0.95      0.90      0.92        20
           3       0.95      0.90      0.92        20

    accuracy                           0.93        54
   macro avg       0.92      0.93      0.93        54
weighted avg       0.93      0.93      0.93        54



[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.0s finished


## 結論:
### PCA的AT&T資料，各個solver('lbfgs','liblinear','newton-cg')表現均差不多，準確率均在90%以上

## 若成分比例採0.6

In [44]:
solver = 'lbfgs'
PCA_LR(n_components = 0.6, solver = solver)

92.59%

              precision    recall  f1-score   support

           1       0.82      1.00      0.90        14
           2       0.95      0.90      0.92        20
           3       1.00      0.90      0.95        20

    accuracy                           0.93        54
   macro avg       0.92      0.93      0.92        54
weighted avg       0.93      0.93      0.93        54



[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.0s finished


## 若成分比例採0.3

In [49]:
solver = 'lbfgs'
PCA_LR(n_components = 0.3, solver = solver)

66.67%

              precision    recall  f1-score   support

           1       0.68      0.93      0.79        14
           2       0.54      0.65      0.59        20
           3       0.91      0.50      0.65        20

    accuracy                           0.67        54
   macro avg       0.71      0.69      0.67        54
weighted avg       0.71      0.67      0.66        54



[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.0s finished


## 結論:
### 可以發現當成分比例採0.8時，準確率都還維持在90%以上，當成分比例採0.4以下時，準確率已經下滑至67%，所以可以推測出當成分比例在0.4左右時，在降低維度的同時也保留原本資料的重要資訊，所以接下來的成分比例都將採0.8跟0.4做比較

## 標準化後原始資料+SVC

## one vs one 

- kernel = 'linear'

In [13]:
from sklearn.svm import SVC,LinearSVC

def clf_SVM(C, opts, clf_svm):
    clf_svm.fit(X_train_,y_train)
    predictions = clf_svm.predict(X_test_)
    print(classification_report(y_test, predictions))

In [14]:
C = 1
opts = dict(C = C, decision_function_shape = 'ovo', \
             tol = 1e-6, max_iter = int(1e6))
clf_svm = SVC(kernel = 'linear', **opts)
clf_SVM(C = C, opts = opts, clf_svm = clf_svm)

              precision    recall  f1-score   support

           1       0.88      1.00      0.93        14
           2       0.95      0.90      0.92        20
           3       1.00      0.95      0.97        20

    accuracy                           0.94        54
   macro avg       0.94      0.95      0.94        54
weighted avg       0.95      0.94      0.94        54



- kernel = 'rbf'

In [15]:
C = 1
opts = dict(C = C, decision_function_shape = 'ovo', \
             tol = 1e-6, max_iter = int(1e6))
clf_svm = SVC(kernel = 'rbf', gamma = 0.2, **opts)
clf_SVM(C = C, opts = opts, clf_svm = clf_svm)

              precision    recall  f1-score   support

           1       1.00      1.00      1.00        14
           2       0.91      1.00      0.95        20
           3       1.00      0.90      0.95        20

    accuracy                           0.96        54
   macro avg       0.97      0.97      0.97        54
weighted avg       0.97      0.96      0.96        54



- kernel = 'poly'

In [16]:
C = 1
opts = dict(C = C, decision_function_shape = 'ovo', \
             tol = 1e-6, max_iter = int(1e6))
clf_svm = SVC(kernel = 'poly', **opts)
clf_SVM(C = C, opts = opts, clf_svm = clf_svm)

              precision    recall  f1-score   support

           1       0.87      0.93      0.90        14
           2       0.76      0.95      0.84        20
           3       1.00      0.70      0.82        20

    accuracy                           0.85        54
   macro avg       0.88      0.86      0.85        54
weighted avg       0.88      0.85      0.85        54



- kernel = 'sigmoid'

In [8]:
C = 1
opts = dict(C=C,tol = 1e-6,max_iter = int(1e6))
clf_svm = SVC(kernel = 'sigmoid', **opts)
clf_svm.fit(X_train_,y_train)
predictions = clf_svm.predict(X_test_)
print(classification_report(y_test, predictions))

              precision    recall  f1-score   support

           1       0.95      1.00      0.97        18
           2       1.00      0.95      0.98        21
           3       1.00      1.00      1.00        15

    accuracy                           0.98        54
   macro avg       0.98      0.98      0.98        54
weighted avg       0.98      0.98      0.98        54



## 結論:
### 在決策函數中選擇一對一(one vs one)的模式，可以看到當 kernel 選擇 'sigmoid' 準確率有最高98%，kernel 選擇 'linear' 或 'rbf' 準確率有90%以上，而 kernel 選擇 'poly' 準確率只有85%，表現得不好

## one vs the rest

In [18]:
C = 1
opts = dict(C = C,tol = 1e-6,max_iter = int(1e6))
clf_svm = LinearSVC(**opts)
clf_SVM(C = C, opts = opts, clf_svm = clf_svm)

              precision    recall  f1-score   support

           1       1.00      1.00      1.00        14
           2       0.95      1.00      0.98        20
           3       1.00      0.95      0.97        20

    accuracy                           0.98        54
   macro avg       0.98      0.98      0.98        54
weighted avg       0.98      0.98      0.98        54



## 結論:
### 1. 當模式為一對其他（one vs the rest），準確率為98%
### 2. 模式為 一對一 跟 一對其他，可以看到準確率皆在90%以上，兩個模式的準確率差不多

## PCA主成分分析(成分比例採0.8) + SVM

## one vs one 

- kernel = 'sigmoid'

In [22]:
from sklearn.decomposition import PCA

pca = PCA(n_components=0.8).fit(X_train_)
Z_train = pca.transform(X_train_) #降維
Z_test = pca.transform(X_test_)
C = 1
opts = dict(C = C, decision_function_shape = 'ovo', \
             tol = 1e-6, max_iter = int(1e6))
clf_svm = SVC(kernel = 'linear',**opts)
clf_svm.fit(Z_train,y_train)
predictions = clf_svm.predict(Z_test)
print(classification_report(y_test, predictions))

              precision    recall  f1-score   support

           1       1.00      1.00      1.00        14
           2       0.90      0.95      0.93        20
           3       0.95      0.90      0.92        20

    accuracy                           0.94        54
   macro avg       0.95      0.95      0.95        54
weighted avg       0.95      0.94      0.94        54



## one vs the rest

In [21]:
from sklearn.decomposition import PCA

pca = PCA(n_components=0.8).fit(X_train_)
Z_train = pca.transform(X_train_) #降維
Z_test = pca.transform(X_test_)
C = 1
opts = dict(C=C,tol = 1e-6,max_iter = int(1e6))
clf_svm = SVC(kernel = 'linear',**opts)
clf_svm.fit(Z_train,y_train)
predictions = clf_svm.predict(Z_test)
print(classification_report(y_test, predictions))

              precision    recall  f1-score   support

           1       1.00      1.00      1.00        14
           2       0.90      0.95      0.93        20
           3       0.95      0.90      0.92        20

    accuracy                           0.94        54
   macro avg       0.95      0.95      0.95        54
weighted avg       0.95      0.94      0.94        54



## 結論:
### 當成分比例採0.8時，不管選擇一對一或一對其他都能維持準確率在90%

## PCA主成分分析(成分比例採0.4) + SVM

## one vs one 

- kernel = 'sigmoid'

In [50]:
from sklearn.decomposition import PCA

pca = PCA(n_components=0.4).fit(X_train_)
Z_train = pca.transform(X_train_) #降維
Z_test = pca.transform(X_test_)
C = 1
opts = dict(C = C, decision_function_shape = 'ovo', \
             tol = 1e-6, max_iter = int(1e6))
clf_svm = SVC(kernel = 'linear',**opts)
clf_svm.fit(Z_train,y_train)
predictions = clf_svm.predict(Z_test)
print(classification_report(y_test, predictions))

              precision    recall  f1-score   support

           1       0.88      1.00      0.93        14
           2       0.95      0.90      0.92        20
           3       1.00      0.95      0.97        20

    accuracy                           0.94        54
   macro avg       0.94      0.95      0.94        54
weighted avg       0.95      0.94      0.94        54



## one vs the rest

In [51]:
from sklearn.decomposition import PCA

pca = PCA(n_components=0.4).fit(X_train_)
Z_train = pca.transform(X_train_) #降維
Z_test = pca.transform(X_test_)
C = 1
opts = dict(C = C, decision_function_shape = 'ovo', \
             tol = 1e-6, max_iter = int(1e6))
clf_svm = SVC(kernel = 'linear',**opts)
clf_svm.fit(Z_train,y_train)
predictions = clf_svm.predict(Z_test)
print(classification_report(y_test, predictions))

              precision    recall  f1-score   support

           1       0.88      1.00      0.93        14
           2       0.95      0.90      0.92        20
           3       1.00      0.95      0.97        20

    accuracy                           0.94        54
   macro avg       0.94      0.95      0.94        54
weighted avg       0.95      0.94      0.94        54



## 結論:
### 當成分比例採0.8或0.4時，不管選擇一對一或一對其他都能維持準確率在90%，因此下面主成分分析的成分比例將採0.4

## 標準化後原始資料+神經網路NN

- hidden_layers = (30,) , activation = 'logistic', solver = 'adam'

In [24]:
from sklearn.neural_network import MLPClassifier

hidden_layers = (30,)
activation = 'logistic'
opts = dict(hidden_layer_sizes = hidden_layers,verbose = True,activation = activation,tol = 1e-6,max_iter = int(1e6))
solver = 'adam'
clf_MLP = MLPClassifier(solver = solver, **opts)
clf_MLP.fit(X_train_,y_train)
predictions = clf_MLP.predict(X_test_)
print(classification_report(y_test,predictions))

Iteration 1, loss = 1.15989271
Iteration 2, loss = 1.14968396
Iteration 3, loss = 1.13975042
Iteration 4, loss = 1.13009480
Iteration 5, loss = 1.12071827
Iteration 6, loss = 1.11162018
Iteration 7, loss = 1.10279781
Iteration 8, loss = 1.09424665
Iteration 9, loss = 1.08596089
Iteration 10, loss = 1.07793380
Iteration 11, loss = 1.07015783
Iteration 12, loss = 1.06262477
Iteration 13, loss = 1.05532580
Iteration 14, loss = 1.04825163
Iteration 15, loss = 1.04139259
Iteration 16, loss = 1.03473868
Iteration 17, loss = 1.02827968
Iteration 18, loss = 1.02200515
Iteration 19, loss = 1.01590456
Iteration 20, loss = 1.00996727
Iteration 21, loss = 1.00418271
Iteration 22, loss = 0.99854034
Iteration 23, loss = 0.99302983
Iteration 24, loss = 0.98764109
Iteration 25, loss = 0.98236436
Iteration 26, loss = 0.97719024
Iteration 27, loss = 0.97210977
Iteration 28, loss = 0.96711446
Iteration 29, loss = 0.96219628
Iteration 30, loss = 0.95734774
Iteration 31, loss = 0.95256181
Iteration 32, los

- hidden_layers = (30,) , activation = 'relu', solver = 'adam'

In [25]:
from sklearn.neural_network import MLPClassifier

hidden_layers = (30,)
activation = 'relu'
opts = dict(hidden_layer_sizes = hidden_layers,verbose = True,activation = activation,tol = 1e-6,max_iter = int(1e6))
solver = 'adam'
clf_MLP = MLPClassifier(solver = solver, **opts)
clf_MLP.fit(X_train_,y_train)
predictions = clf_MLP.predict(X_test_)
print(classification_report(y_test,predictions))

Iteration 1, loss = 1.61379971
Iteration 2, loss = 1.58443495
Iteration 3, loss = 1.55547224
Iteration 4, loss = 1.52710410
Iteration 5, loss = 1.49917418
Iteration 6, loss = 1.47166037
Iteration 7, loss = 1.44459367
Iteration 8, loss = 1.41793431
Iteration 9, loss = 1.39169964
Iteration 10, loss = 1.36597658
Iteration 11, loss = 1.34079763
Iteration 12, loss = 1.31608679
Iteration 13, loss = 1.29179000
Iteration 14, loss = 1.26783615
Iteration 15, loss = 1.24434067
Iteration 16, loss = 1.22125097
Iteration 17, loss = 1.19864769
Iteration 18, loss = 1.17654330
Iteration 19, loss = 1.15488640
Iteration 20, loss = 1.13361936
Iteration 21, loss = 1.11272568
Iteration 22, loss = 1.09217938
Iteration 23, loss = 1.07203572
Iteration 24, loss = 1.05233570
Iteration 25, loss = 1.03295441
Iteration 26, loss = 1.01393830
Iteration 27, loss = 0.99532049
Iteration 28, loss = 0.97710473
Iteration 29, loss = 0.95925667
Iteration 30, loss = 0.94174553
Iteration 31, loss = 0.92452875
Iteration 32, los

- hidden_layers = (512,)

In [27]:
from sklearn.neural_network import MLPClassifier

hidden_layers = (512,)
activation = 'logistic'
opts = dict(hidden_layer_sizes = hidden_layers,verbose = True,activation = activation,tol = 1e-6,max_iter = int(1e6))
solver = 'adam'
clf_MLP = MLPClassifier(solver = solver, **opts)
clf_MLP.fit(X_train_,y_train)
predictions = clf_MLP.predict(X_test_)
print(classification_report(y_test,predictions))

Iteration 1, loss = 1.09985299
Iteration 2, loss = 1.04282847
Iteration 3, loss = 1.02569089
Iteration 4, loss = 1.01622552
Iteration 5, loss = 0.99830429
Iteration 6, loss = 0.97295542
Iteration 7, loss = 0.94632683
Iteration 8, loss = 0.92303893
Iteration 9, loss = 0.90448078
Iteration 10, loss = 0.88871232
Iteration 11, loss = 0.87226694
Iteration 12, loss = 0.85313248
Iteration 13, loss = 0.83193468
Iteration 14, loss = 0.81064671
Iteration 15, loss = 0.79082802
Iteration 16, loss = 0.77278823
Iteration 17, loss = 0.75578367
Iteration 18, loss = 0.73874194
Iteration 19, loss = 0.72095376
Iteration 20, loss = 0.70238167
Iteration 21, loss = 0.68352619
Iteration 22, loss = 0.66503578
Iteration 23, loss = 0.64732052
Iteration 24, loss = 0.63036529
Iteration 25, loss = 0.61381617
Iteration 26, loss = 0.59725970
Iteration 27, loss = 0.58049385
Iteration 28, loss = 0.56361418
Iteration 29, loss = 0.54689800
Iteration 30, loss = 0.53060865
Iteration 31, loss = 0.51485860
Iteration 32, los

訓練資料的classification report 

In [28]:
pred = clf_MLP.predict(X_train_)
print(classification_report(y_train,pred))

              precision    recall  f1-score   support

           1       1.00      1.00      1.00        45
           2       1.00      1.00      1.00        51
           3       1.00      1.00      1.00        28

    accuracy                           1.00       124
   macro avg       1.00      1.00      1.00       124
weighted avg       1.00      1.00      1.00       124



- hidden_layers = (60, 60, 60)

In [29]:
from sklearn.neural_network import MLPClassifier

hidden_layers = (60,60,60)
activation = 'logistic'
opts = dict(hidden_layer_sizes = hidden_layers,verbose = True,activation = activation,tol = 1e-6,max_iter = int(1e6))
solver = 'adam'
clf_MLP = MLPClassifier(solver = solver, **opts)
clf_MLP.fit(X_train_,y_train)
predictions = clf_MLP.predict(X_test_)
print(classification_report(y_test,predictions))

Iteration 1, loss = 1.08756968
Iteration 2, loss = 1.07820776
Iteration 3, loss = 1.07212742
Iteration 4, loss = 1.06915770
Iteration 5, loss = 1.06869760
Iteration 6, loss = 1.06964583
Iteration 7, loss = 1.07073085
Iteration 8, loss = 1.07109747
Iteration 9, loss = 1.07053781
Iteration 10, loss = 1.06928073
Iteration 11, loss = 1.06770834
Iteration 12, loss = 1.06617760
Iteration 13, loss = 1.06493379
Iteration 14, loss = 1.06407729
Iteration 15, loss = 1.06356735
Iteration 16, loss = 1.06325890
Iteration 17, loss = 1.06296406
Iteration 18, loss = 1.06251783
Iteration 19, loss = 1.06182368
Iteration 20, loss = 1.06086634
Iteration 21, loss = 1.05969593
Iteration 22, loss = 1.05839679
Iteration 23, loss = 1.05705406
Iteration 24, loss = 1.05572679
Iteration 25, loss = 1.05443241
Iteration 26, loss = 1.05314502
Iteration 27, loss = 1.05180665
Iteration 28, loss = 1.05034732
Iteration 29, loss = 1.04870619
Iteration 30, loss = 1.04684629
Iteration 31, loss = 1.04475816
Iteration 32, los

- hidden_layers = (512, 512, 512)

In [30]:
from sklearn.neural_network import MLPClassifier

hidden_layers = (512,512,512)
activation = 'logistic'
opts = dict(hidden_layer_sizes = hidden_layers,verbose = True,activation = activation,tol = 1e-6,max_iter = int(1e6))
solver = 'adam'
clf_MLP = MLPClassifier(solver = solver, **opts)
clf_MLP.fit(X_train_,y_train)
predictions = clf_MLP.predict(X_test_)
print(classification_report(y_test,predictions))

Iteration 1, loss = 1.22132922
Iteration 2, loss = 1.09902323
Iteration 3, loss = 1.20627408
Iteration 4, loss = 1.14754799
Iteration 5, loss = 1.09991528
Iteration 6, loss = 1.08204642
Iteration 7, loss = 1.08576131
Iteration 8, loss = 1.10617276
Iteration 9, loss = 1.11062644
Iteration 10, loss = 1.09218677
Iteration 11, loss = 1.07107543
Iteration 12, loss = 1.06267834
Iteration 13, loss = 1.06733741
Iteration 14, loss = 1.07620278
Iteration 15, loss = 1.08054425
Iteration 16, loss = 1.07745992
Iteration 17, loss = 1.06942146
Iteration 18, loss = 1.06078627
Iteration 19, loss = 1.05494924
Iteration 20, loss = 1.05299842
Iteration 21, loss = 1.05361053
Iteration 22, loss = 1.05411373
Iteration 23, loss = 1.05224739
Iteration 24, loss = 1.04736614
Iteration 25, loss = 1.04029506
Iteration 26, loss = 1.03240271
Iteration 27, loss = 1.02482320
Iteration 28, loss = 1.01794448
Iteration 29, loss = 1.01096275
Iteration 30, loss = 1.00202908
Iteration 31, loss = 0.98948517
Iteration 32, los

訓練資料的classification report

In [31]:
pred = clf_MLP.predict(X_train_)
print(classification_report(y_train,pred))

              precision    recall  f1-score   support

           1       1.00      1.00      1.00        45
           2       1.00      1.00      1.00        51
           3       1.00      1.00      1.00        28

    accuracy                           1.00       124
   macro avg       1.00      1.00      1.00       124
weighted avg       1.00      1.00      1.00       124



## 結論:
### 1. hidden layers = (30,)，選擇不同的activation 跟 solver ，，因為資料量很小，效果並沒有差太多，準確率皆在90%以上
### 2. 當hidden layers = (512,) 跟 hidden layers = (60,60,60) 或 hidden layers = (512,512,512)，差別並沒有太大，只用一層512個神經元就可以達到跟三層60個神經元或是三層512個神經元的效果，所以下面採hidden layers = (512,)
### 3. 由2.可以知道，低維度的資料如果擴展到過多的神經元，其實資料無法提供足夠多的細節，也因此造成了許多無用、重複的數據。有時候不是模型大、模型深就是會比較好，要根據不同的狀況做適當的設計才是最好的選擇

## PCA主成分分析(成分比例採0.4)+神經網路NN

- hidden_layers = (30,)

In [52]:
from sklearn.decomposition import PCA

hidden_layers = (30,)
activation = 'logistic'
opts = dict(hidden_layer_sizes = hidden_layers,verbose = True,activation = activation,tol = 1e-6,max_iter = int(1e6))
solver = 'adam'
clf_MLP = MLPClassifier(solver = solver, **opts)
clf_MLP.fit(Z_train,y_train)
predictions = clf_MLP.predict(Z_test)
print(classification_report(y_test,predictions))

Iteration 1, loss = 0.98988869
Iteration 2, loss = 0.98629936
Iteration 3, loss = 0.98271912
Iteration 4, loss = 0.97914649
Iteration 5, loss = 0.97557874
Iteration 6, loss = 0.97201442
Iteration 7, loss = 0.96845310
Iteration 8, loss = 0.96489461
Iteration 9, loss = 0.96133883
Iteration 10, loss = 0.95778564
Iteration 11, loss = 0.95423495
Iteration 12, loss = 0.95068665
Iteration 13, loss = 0.94714067
Iteration 14, loss = 0.94359693
Iteration 15, loss = 0.94005534
Iteration 16, loss = 0.93651584
Iteration 17, loss = 0.93297834
Iteration 18, loss = 0.92944275
Iteration 19, loss = 0.92590898
Iteration 20, loss = 0.92237694
Iteration 21, loss = 0.91884651
Iteration 22, loss = 0.91531758
Iteration 23, loss = 0.91179003
Iteration 24, loss = 0.90826373
Iteration 25, loss = 0.90473854
Iteration 26, loss = 0.90121434
Iteration 27, loss = 0.89769099
Iteration 28, loss = 0.89416835
Iteration 29, loss = 0.89064630
Iteration 30, loss = 0.88712472
Iteration 31, loss = 0.88360346
Iteration 32, los

In [53]:
pred = clf_MLP.predict(Z_train)
print(classification_report(y_train,pred))

              precision    recall  f1-score   support

           1       1.00      1.00      1.00        45
           2       1.00      1.00      1.00        51
           3       1.00      1.00      1.00        28

    accuracy                           1.00       124
   macro avg       1.00      1.00      1.00       124
weighted avg       1.00      1.00      1.00       124



- hidden_layers = (512,)

In [55]:
from sklearn.decomposition import PCA

hidden_layers = (512,)
activation = 'logistic'
opts = dict(hidden_layer_sizes = hidden_layers,verbose = True,activation = activation,tol = 1e-6,max_iter = int(1e6))
solver = 'adam'
clf_MLP = MLPClassifier(solver = solver, **opts)
clf_MLP.fit(Z_train,y_train)
predictions = clf_MLP.predict(Z_test)
print(classification_report(y_test,predictions))

Iteration 1, loss = 1.18123021
Iteration 2, loss = 1.10115075
Iteration 3, loss = 1.07781314
Iteration 4, loss = 1.08883961
Iteration 5, loss = 1.09155232
Iteration 6, loss = 1.07651814
Iteration 7, loss = 1.05448096
Iteration 8, loss = 1.03605757
Iteration 9, loss = 1.02547738
Iteration 10, loss = 1.02043284
Iteration 11, loss = 1.01604903
Iteration 12, loss = 1.00892246
Iteration 13, loss = 0.99830569
Iteration 14, loss = 0.98553558
Iteration 15, loss = 0.97282845
Iteration 16, loss = 0.96193745
Iteration 17, loss = 0.95324753
Iteration 18, loss = 0.94576825
Iteration 19, loss = 0.93800071
Iteration 20, loss = 0.92900820
Iteration 21, loss = 0.91886992
Iteration 22, loss = 0.90832363
Iteration 23, loss = 0.89808829
Iteration 24, loss = 0.88845313
Iteration 25, loss = 0.87930972
Iteration 26, loss = 0.87037870
Iteration 27, loss = 0.86136669
Iteration 28, loss = 0.85204852
Iteration 29, loss = 0.84235475
Iteration 30, loss = 0.83241469
Iteration 31, loss = 0.82247258
Iteration 32, los

In [56]:
pred = clf_MLP.predict(Z_train)
print(classification_report(y_train,pred))

              precision    recall  f1-score   support

           1       1.00      1.00      1.00        45
           2       1.00      1.00      1.00        51
           3       1.00      1.00      1.00        28

    accuracy                           1.00       124
   macro avg       1.00      1.00      1.00       124
weighted avg       1.00      1.00      1.00       124



## 結論:
### 1. 當成分比例採0.4時，準確率還不錯，為94%
### 2. hidden layers = (30,) 跟 hidden layers = (512,) PCA的結果差不多，所以可以選擇一層30個神經元即可